lib/Conversion/VerifToSV/VerifToSV.cpp

```diff
@@ -0,0 +1,78 @@
+//===- VerifToSV.cpp - HW To SV Conversion Pass ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This is the main Verif to SV Conversion Pass Implementation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/VerifToSV.h"
+#include "../PassDetail.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace sv;
+using namespace verif;
+
+namespace {
+struct VerifToSVPass : public LowerVerifToSVBase<VerifToSVPass> {
+  void runOnOperation() override;
+};
+
+struct PrintfOpConversionPattern : public OpConversionPattern<PrintfOp> {
+  using OpConversionPattern<PrintfOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(PrintfOp op, OpAdaptor operands,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    // Printf's will be emitted to stdout (32'h8000_0001 in IEEE Std 1800-2012).
+    Value fdStdout = rewriter.create<hw::ConstantOp>(
+        op.getLoc(), APInt(32, 0x80000001, false));
+
+    auto fstrOp = dyn_cast_or_null<FormatStringOp>(
+        op.getFormattedString().getDefiningOp());
+    if (!fstrOp)
+      return op->emitOpError()
+             << "expected FormatStringOp as the source of the formatted string";
```

Comment:
This looks like a good starting point! In the future we might have something like `hw.constant_string` which we could accept here as well. Or instead, absorbing `FormatStringOp` could be an optimized lowering and the default simple lowering would just be the `!hw.string` being passed into the `sv::FWriteOp` as is.

Reply:
I agree -- `hw.string` should be converted to a sv string (in export verilog) and the default lowering for fmtstring should be `$*format` to make this more composable.